fetch/mesh_data.go

```diff
@@ -353,49 +319,33 @@ func (f *Fetch) GetCert(
 		Layer: lid,
 		Block: &bid,
 	}
-	reqData, err := codec.Encode(req)
-	if err != nil {
-		f.logger.With().Fatal("failed to encode cert request", log.Err(err))
-	}
+	reqData := codec.MustEncode(req)
 
-	out := make(chan *types.Certificate, 1)
 	for _, peer := range peers {
-		done := make(chan error, 1)
-		okCB := func(data []byte) {
-			var peerCert types.Certificate
-			if err = codec.Decode(data, &peerCert); err != nil {
-				done <- err
-				return
-			}
-			// for generic data fetches by hash (ID for atx/block/proposal/ballot/tx), the check on whether the returned
-			// data matching the hash was done on the data handlers' path. for block certificate, there is no ID associated
-			// with it, hence the check here.
-			// however, certificate doesn't go through that path. it's requested by a separate protocol because a block
-			// certificate doesn't have an ID.
-			if peerCert.BlockID != bid {
-				done <- fmt.Errorf("peer %v served wrong cert. want %s got %s", peer, bid.String(), peerCert.BlockID.String())
-				return
-			}
-			out <- &peerCert
-		}
-		errCB := func(perr error) {
-			done <- perr
+		data, err := f.servers[OpnProtocol].Request(ctx, peer, reqData)
+		if err != nil {
+			f.logger.With().Debug("failed to get cert", log.Stringer("peer", peer), log.Err(err))
+			continue
 		}
-		if err := f.servers[OpnProtocol].Request(ctx, peer, reqData, okCB, errCB); err != nil {
-			done <- err
+		var peerCert types.Certificate
+		if err = codec.Decode(data, &peerCert); err != nil {
+			f.logger.With().Debug("failed to decode cert", log.Stringer("peer", peer), log.Err(err))
+			continue
 		}
-		select {
-		case err := <-done:
-			f.logger.With().Debug("failed to get cert from peer",
+		// for generic data fetches by hash (ID for atx/block/proposal/ballot/tx), the check on whether the returned
+		// data matching the hash was done on the data handlers' path. for block certificate, there is no ID associated
+		// with it, hence the check here.
+		// however, certificate doesn't go through that path. it's requested by a separate protocol because a block
+		// certificate doesn't have an ID.
+		if peerCert.BlockID != bid {
+			f.logger.With().Debug(
+				"peer served wrong cert",
+				log.Stringer("want", bid),
+				log.Stringer("got", peerCert.BlockID),
 				log.Stringer("peer", peer),
-				log.Err(err),
 			)
-			continue
-		case cert := <-out:
-			return cert, nil
-		case <-ctx.Done():
-			return nil, ctx.Err()
 		}
+		return &peerCert, nil
 	}
-	return nil, fmt.Errorf("failed to get cert %v/%s from %d peers", lid, bid.String(), len(peers))
+	return nil, fmt.Errorf("failed to get cert %v/%s from %d peers: %w", lid, bid.String(), len(peers), ctx.Err())
```

Comment:
The previous code would send requests to all peers simultaneously, process the first it receives successfully and return. The new code queries each other peer sequentially. How about something like this:

```go
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()
	res := make(chan *types.Certificate, 1)
	var eg errgroup.Group
	for i := range peers {
		peer := peers[i]

		eg.Go(func() error {
			data, err := f.servers[OpnProtocol].Request(ctx, peer, reqData)
			if err != nil {
				f.logger.With().Debug("failed to get cert", log.Stringer("peer", peer), log.Err(err))
				return nil
			}
			var peerCert types.Certificate
			if err = codec.Decode(data, &peerCert); err != nil {
				f.logger.With().Debug("failed to decode cert", log.Stringer("peer", peer), log.Err(err))
				return nil
			}
			// for generic data fetches by hash (ID for atx/block/proposal/ballot/tx), the check on whether the returned
			// data matching the hash was done on the data handlers' path. for block certificate, there is no ID associated
			// with it, hence the check here.
			// however, certificate doesn't go through that path. it's requested by a separate protocol because a block
			// certificate doesn't have an ID.
			if peerCert.BlockID != bid {
				f.logger.With().Debug(
					"peer served wrong cert",
					log.Stringer("want", bid),
					log.Stringer("got", peerCert.BlockID),
					log.Stringer("peer", peer),
				)
			}
			select {
			case <-ctx.Done():
				return ctx.Err()
			case res <- &peerCert:
				return nil
			}
		})
	}
	select {
	case <-ctx.Done():
		return nil, fmt.Errorf("failed to get cert %v/%s from %d peers: %w", lid, bid.String(), len(peers), ctx.Err())
	case cert := <-res:
		cancel()
		eg.Wait()
		return cert, nil
	}
```

This would also query all peers simultaneously, process the first certificate it receives and then cancel all other requests before returning. The old code also had the problem that requests were not canceled after receiving a successful response, while this code should do that.

Reply:
Are you sure? I think the original code queried peers one by one.

Reply:
Since all requests are canceled as soon as we receive the first certificate it probably makes sense to not do this concurrently 🤔

Reply:
Only one request at a time was sent before because directly under the call to `Request()` is a `select` waiting for a response (or error).

Reply:
You right, I missed that the select was part of the loop.